packages/ui/src/components/select/select.tsx

```diff
@@ -0,0 +1,570 @@
+import * as React from "react"
+import {
+  UseComboboxProps,
+  UseComboboxReturnValue,
+  UseComboboxState,
+  UseComboboxStateChange,
+  UseComboboxStateChangeOptions,
+  useCombobox,
+} from "downshift"
+import { isEqual } from "lodash"
+import {
+  CheckMini,
+  ChevronUpDown,
+  EllipseMiniSolid,
+  MagnifyingGlassMini,
+  XMarkMini,
+} from "@medusajs/icons"
+import { DropdownMenu } from "@/components/dropdown-menu"
+import { clx } from "@/utils/clx"
+import { Badge } from "../badge"
+import { labelVariants } from "../label"
+
+const ALLOWED_SEARCH_KEYDOWN_CODES = ["Enter", "Escape", "ArrowUp", "ArrowDown"]
+const SCROLL_TOLERANCE = 30
+
+type SelectState<T> =
+  | ({
+      multi: boolean
+      selectedItems?: SelectItem[]
+      clearSelectedItems: () => void
+      selectAll: () => void
+      allSelected: boolean
+      search: boolean
+      onSearch: (searchTerm: string) => void
+      addItem: (item: SelectItem) => void
+      removeItem: (item: SelectItem) => void
+      onScrollToBottom: () => void
+    } & UseComboboxReturnValue<T>)
+  | null
```

Comment:
Not sure why this is a generic when it's forced to use SelectItem in the context definition, and if you want to keep it generic shouldn't it be:

```suggestion
type SelectState<T> = {
      multi: boolean
      selectedItems?: T[]
      clearSelectedItems: () => void
      selectAll: () => void
      allSelected: boolean
      search: boolean
      onSearch: (searchTerm: string) => void
      addItem: (item: T) => void
      removeItem: (item: T) => void
      onScrollToBottom: () => void
    } & UseComboboxReturnValue<T>
```